salt/modules/junos.py

```diff
@@ -131,8 +131,9 @@ def __enter__(self):
                 "Caching file {0} at {1}".format(self._file_path, self._cached_folder)
             )
             if self._kwargs:
-                self._cached_file = __salt__["cp.get_template"](
-                    self._file_path, self._cached_folder, **self._kwargs
+                self._cached_file = salt.utils.files.mkstemp(dir=self._cached_folder)
```

Comment:
Wondering why the mkdirs=True parameter is not being used. See https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.cp.html#salt.modules.cp.get_template

Reply:
As we need to mandatorily pass `dest` location to cp.get_template https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.cp.html#salt.modules.cp.get_template
The temporary dest (which gets cleaned after API work is done) is already created using https://github.com/saltstack/salt/blob/af051441aa70f8caa234e51bf423a71c8d66be8b/salt/modules/junos.py#L129 hence we don't need to pass mkdirs.

Reply:
@vnitinv My comment 'makedirs=True' is related to why that is not the proper fix for this, not 'well I am using mkdtemp'.  When there is already a solution for having directories created, it should be used, unless there is problem using 'mkdirs=True'.  Already have code using mkdtemp in the PR is not a relevant answer to a review question.

BTW this will not make it in the Sodium release as that is tagged, packaged, tested and QA and about to be released tomorrow. Salt point release 3000.1 is the earliest release that this PR could be applied to.

Reply:
@dmurphy18, I am missing something it seems. 

I could just pass dir=/tmp and makedirs=True but that giving me error everytime (even if I delete the file it breaks for, it error for another file). This is irrespective of we use makedirs or not.

root@iagent:/# salt proxy01 cp.get_template salt://configs/pankaj.set /tmp makedirs=True
```
proxy01:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/salt/metaproxy/proxy.py", line 480, in thread_return
        opts, data, func, args, kwargs
      File "/usr/local/lib/python3.6/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/salt/modules/cp.py", line 327, in get_template
        return _client().get_template(path, dest, template, makedirs, saltenv, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/salt/fileclient.py", line 830, in get_template
        shutil.move(data["data"], dest)
      File "/usr/lib/python3.6/shutil.py", line 548, in move
        raise Error("Destination path '%s' already exists" % real_dst)
    shutil.Error: Destination path '/tmp/__salt.tmp.vgfq0pwc' already exists
ERROR: Minions returned with non-zero exit code
root@iagent:/# rm /tmp/__salt.tmp.vgfq0pwc
root@iagent:/# salt proxy01 cp.get_template salt://configs/pankaj.set /tmp makedirs=True
proxy01:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/salt/metaproxy/proxy.py", line 480, in thread_return
        opts, data, func, args, kwargs
      File "/usr/local/lib/python3.6/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/salt/modules/cp.py", line 327, in get_template
        return _client().get_template(path, dest, template, makedirs, saltenv, **kwargs)
      File "/usr/local/lib/python3.6/dist-packages/salt/fileclient.py", line 830, in get_template
        shutil.move(data["data"], dest)
      File "/usr/lib/python3.6/shutil.py", line 548, in move
        raise Error("Destination path '%s' already exists" % real_dst)
    shutil.Error: Destination path '/tmp/__salt.tmp.ljdtvx6a' already exists
ERROR: Minions returned with non-zero exit code
```

Anyways, **I have changed the logic (with the last commit) and now don't use any dirs things for cp.get_template, and use tempfile itself. I hope this is the best we should have.
Let me know if we are good now.**

I am OK with this change going to whichever latest build it can (if not sodium)

Reply:
Added @dwoz as reviewer

Reply:
@vnitinv Sorry for taking so long to get back to you on this, looks good